### PR TITLE
Clarify on non-Module roots

### PR DESCRIPTION
**Pull request by temyurchenko**
The PR fixes a missing parent on the "__main__" node and weakens the typing of "NodeNG.root" to better reflect the reality

I don't know what entry to fill in in the ChangeLog.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :scroll: Docs          |

## Description

Refs #120, partially.

### CODE REVIEW

**Comment by temyurchenko** [comment]
The reason the CI is failing is that at the class construction stage, it has "Unknown" for a parent. On the previous revision, the result was just that all builtin (or not just?) classes have "Unknown" as their `locals['__module__']`.
I'm thinking of a better way to approach it in this version.

**Comment by DanielNoord** [comment]
I think this will cause a lot of type errors in `pylint` as we assume `root()` always returns... Can we get away by asserting?

**Comment by temyurchenko** [comment]
> I think this will cause a lot of type errors in `pylint` as we assume `root()` always returns... Can we get away by asserting?

I'm currently in the process of eliminating non-module roots. Without that, the assert is going to fail a loot of times.

**Comment by temyurchenko** [comment]
> I think this will cause a lot of type errors in `pylint` as we assume `root()` always returns... Can we get away by asserting?

What is the difference between `Uninferable` and `InferenceError`? Which one should be use and in what cases? Is there some documentation on it? Or a specific person to ask?

**Comment by DanielNoord** [comment]
> > I think this will cause a lot of type errors in `pylint` as we assume `root()` always returns... Can we get away by asserting?
> 
> I'm currently in the process of eliminating non-module roots. Without that, the assert is going to fail a loot of times.

I don't understand what you mean with this?

> > I think this will cause a lot of type errors in `pylint` as we assume `root()` always returns... Can we get away by asserting?
> 
> What is the difference between `Uninferable` and `InferenceError`? Which one should be used and in what cases? Is there some documentation on it? Or a specific person to ask?

It depends on the situation. Ideally `InferenceError` isn't exposed to the users of `astroid` such as `pylint` so that they only need to handle `Uninferable`.

**Comment by temyurchenko** [comment]
> I don't understand what you mean with this?

The `root()` currently returns non-Module nodes, often. So, if we just assert that the return is a Module, that assert is going to fail often.

> It depends on the situation. Ideally `InferenceError` isn't exposed to the users of `astroid` such as `pylint` so that they only need to handle `Uninferable`.

What is the logic for preferring one or the other internally? I often see lines like:

```
            if not name.startswith("__") and self.has_dynamic_getattr(context):
                # class handle some dynamic attributes, return a Uninferable object
                yield util.Uninferable
            else:
                raise InferenceError(
                    str(error), target=self, attribute=name, context=context
                ) from error
```
(this one is from `scoped_nodes.ClassDef.igetattr`)

**Comment by temyurchenko** [comment]
> > I don't understand what you mean with this?
> 
> The `root()` currently returns non-Module nodes, often. So, if we just assert that the return is a Module, that assert is going to fail often.
> 
> > It depends on the situation. Ideally `InferenceError` isn't exposed to the users of `astroid` such as `pylint` so that they only need to handle `Uninferable`.
> 
> What is the logic for preferring one or the other internally? I often see lines like:
> 
> ```
>             if not name.startswith("__") and self.has_dynamic_getattr(context):
>                 # class handle some dynamic attributes, return a Uninferable object
>                 yield util.Uninferable
>             else:
>                 raise InferenceError(
>                     str(error), target=self, attribute=name, context=context
>                 ) from error
> ```
> 
> (this one is from `scoped_nodes.ClassDef.igetattr`)



> What is the logic for preferring one or the other internally? I often see lines like:

I see, raising the InferenceError doesn't let adding new inference results, while just yielding uninferrable does allow it.

**Comment by temyurchenko** [comment]
Now, this should also close #1490. 

I tried my best to give complete descriptions to the commits, because they are lengthy. I also changed the annotation of `root()` back to just `Module` (with an added assert). 

I'm happy to elaborate more on changes, since there are quite a few of them.

As a side note, I still feel that it's not a very honest type signature, given how many nodes we are creating out-of-tree and thus without a parent (see `const_factory`). Furthermore, there is no enforcement of the `parent is not None` invariant anywhere in the codebase, so the accidents are bound to creep in.

**Comment by temyurchenko** [comment]
Also fixes #2517

File: astroid/bases.py
Comment (DanielNoord) [inline astroid/bases.py:8]: Please remove this line

File: astroid/bases.py
Comment (DanielNoord) [inline astroid/bases.py:278]: Could you split out these changes into a separate commit/PR? I'd like to see the effect of these changes on the tests by itself.

File: astroid/brain/brain_argparse.py
Comment (DanielNoord) [inline astroid/brain/brain_argparse.py:23]: Not sure about the name of `adhoc`. But I also don't have better suggestions

File: astroid/brain/brain_builtin_inference.py
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:7]: Please remove this line

File: astroid/brain/brain_builtin_inference.py
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:649]: Why this change? Doesn't that lose value?

File: astroid/brain/brain_builtin_inference.py
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:654]: I don't understand this change? Why doesn't the property have the parent of the function being decorated as its parent?

File: astroid/brain/brain_namedtuple_enum.py
Comment (DanielNoord) [inline astroid/brain/brain_namedtuple_enum.py:154]: I think L156 can be remove?

File: astroid/interpreter/objectmodel.py
Comment (DanielNoord) [inline astroid/interpreter/objectmodel.py:24]: Not going to comment on all, but please remove these lines

**Review by DanielNoord — COMMENTED** [review]
Awesome changes! But also pretty hard to review.

Maintainer time on `astroid` is sparse and nobody has actively worked on its internals for about a year.

I'd be open to spending some time of the next days/weeks to get this merged but I an only approve smaller bite sized PRs. If you're op for that let's go down that route!

For example, the first commit could be a separate PR to declutter this PR.
The second commit does a lot of things at once (as indicated by the commit message). I think a lot of changes in there could be separated into separate PRs, also to show their effects on the tests more clearly. That helps with reviewing them.

**Comment by temyurchenko** [comment]
> I'd be open to spending some time of the next days/weeks to get this merged but I an only approve smaller bite sized PRs. If you're op for that let's go down that route!

For sure.

> For example, the first commit could be a separate PR to declutter this PR.

Will do.

> The second commit does a lot of things at once (as indicated by the commit message). I think a lot of changes in there could be separated into separate PRs, also to show their effects on the tests more clearly. That helps with reviewing them.

Will do.

Regarding isort. I've grepped the project and found a few similar isort comment. Based on that I concluded that the project uses isort to sort imports. However, when I ran isort on it, it resulted in quite a few changes, thus the "isort: off" comments. Could you please tell me if I should use isort? If so, is there a special config for the project? I wasn't able to find one.

File: astroid/brain/brain_argparse.py
Comment (temyurchenko) [reply astroid/brain/brain_argparse.py:23]: I can come up with other options. Please, tell what you are unsure about it, to guide the thought process for new names.

File: astroid/brain/brain_builtin_inference.py
Comment (temyurchenko) [reply astroid/brain/brain_builtin_inference.py:649]: Right, I talk about it in the commit message:

- fixing construction of in-place properties (`bar =
   property(getter)`). They just create a nameless object, not the one
   with the name of the getter. Thus, the name was changed to
   "<property>". Furthermore, the definition of that property is not
   attached to any scope, as it's again nameless.

This change is for "inplace", nameless, properties, not the ones created with a decorator. 

What does it mean for it to lose the value? It's still accessible through an `Assign` field, for example.

File: astroid/brain/brain_builtin_inference.py
Comment (temyurchenko) [reply astroid/brain/brain_builtin_inference.py:654]: See the comment above, this is about «inplace» properties, not the «decorator» properties.

**Comment by DanielNoord** [comment]
https://github.com/pylint-dev/astroid/blob/7954bac2f4e84e3f0b1096be246b238a5651c76e/pyproject.toml#L146

We use `ruff` to do our import sorting now. Any comments you found are probably outdated (and could be removed in a separate PR)

File: astroid/brain/brain_builtin_inference.py
Comment (DanielNoord) [reply astroid/brain/brain_builtin_inference.py:649]: Ah, makes sense. I understand the name change a little better now (although I would probably do `f"<property> {inferred.name}"` or something, so we don't lose that context.

I don't really understand the `parent` changes though.

```python
class A:
    temperature = property(get_temperature, set_temperature)
```

For `temperature` I would still expect `A` to be the `parent` no? Not the `adhoc` module

File: astroid/brain/brain_argparse.py
Comment (DanielNoord) [reply astroid/brain/brain_argparse.py:23]: @Pierre-Sassoulas Any suggestions here?

File: astroid/brain/brain_builtin_inference.py
Comment (temyurchenko) [reply astroid/brain/brain_builtin_inference.py:649]: For `temperature` it is. However, temperature is just a name in the AssignName. It is assigned a call to the constructor of actual `property`. The actual `property` class is defined somewhere in the builtins module. However, since we are treating `property` specially, we are creating our own definition of it, separate for each use. Kind of like an Instance. And that special definition isn't defined anywhere. We've just made it up on the spot, ad-hoc. 

Treating it as if it was defined in `A` is furthermore incorrect, since it would cause it to be in `A`s body, or searchable from `A`.

**Comment by temyurchenko** [comment]
> I think a lot of changes in there could be separated into separate PRs, also to show their effects on the tests more clearly.

Is it possible to have something like an merge train on github? That is, for one PR to depend on some previous PR? It's possible to emulate with merging into branches, but then you'd have to do reviews on my local fork, since I can't create branches in this repository.

**Comment by DanielNoord** [comment]
> > I think a lot of changes in there could be separated into separate PRs, also to show their effects on the tests more clearly.
> 
> 
> 
> Is it possible to have something like an merge train on github? That is, for one PR to depend on some previous PR? It's possible to emulate with merging into branches, but then you'd have to do reviews on my local fork, since I can't create branches in this repository.
> 
> 
> 
> If not, the simplest way would probably be do just breakup commits within this PR. Would you be open to that?

Sadly it doesn't. 

I'm okay with splitting up in this PR but I think that increases the risk of me (and other maintainers) finding it hard to approve of the full PR and merge it.

Separate PRs make it much easier to spend 15-20 minutes during lunch to review and merge the smaller commits. From experience I know although it involves more rebase work in the end it significantly increases the chances of `astroid` PRs landing on `main`.

**Comment by temyurchenko** [comment]
> I'm okay with splitting up in this PR but I think that increases the risk of me (and other maintainers) finding it hard to approve of the full PR and merge it.

Sure, I'll do my best to achieve that.

**Comment by temyurchenko** [comment]
I've made the first batch of PRs that cover a big part of this one. One thing is, #2557 is better reviewed the last.

**Comment by Pierre-Sassoulas** [comment]
The secret to merge PR with volunteer maintainers has been told. I have multiple 5mn free windows on mobile during the days. I'll hit that merge button when my rust code is compiling, when the pasta's water is heating, when I'm waiting for the train, or when a butterfly distracted me and I checked the notification on my phone. I have some 10mn/20mn windows during the week. But one hour of uninterrupted concentration time on PC ? I have to *plan* ahead for that, it's not going to happens (not in a reasonable time at least)

**Comment by temyurchenko** [comment]
> The secret to merge PR with volunteer maintainers has been told. I have multiple 5mn free windows on mobile during the days. I'll hit that merge button when my rust code is compiling, when the pasta's water is heating, when I'm waiting for the train, or when a butterfly distracted me and I checked the notification on my phone. I have some 10mn/20mn windows during the week. But one hour of uninterrupted concentration time on PC ? I have to _plan_ ahead for that, it's not going to happens (not in a reasonable time at least)

I did my best to split off the PR into smaller ones, as you can see in the comment above. They currently don't cover the whole of this PR, but I will finish the job some time later.

File: astroid/brain/brain_builtin_inference.py
Comment (DanielNoord) [reply astroid/brain/brain_builtin_inference.py:648]: I understand now, thanks!

**Comment by DanielNoord** [comment]
I think I went through all PRs and merged the first one that should solve most rebase issues. Let me know when you rebased the others or need me to do another review!

**Comment by DanielNoord** [comment]
To give an update: changes coming a long nicely. @temyurchenko has been very helpful in rebasing and responding to comments. 

https://github.com/pylint-dev/astroid/pull/2563 is on my radar, but after a long week at work I don't really have the capacity to have a good look at it. I hope to do so this weekend or on Monday.

**Review by DanielNoord — DISMISSED** [review]
Awesome! Great job working with us to get this merged. Working with you was a joy!

_I would be okay with adding a Changelog for this as I think you work should be recognized!_

File: astroid/brain/brain_argparse.py
Comment (Pierre-Sassoulas) [reply astroid/brain/brain_argparse.py]: Maybe ``synthetic_root_node``, ``placeholder_root_node``,  ``placeholder_parent_node``,  ``placeholder_module_node`` ?  wasn't sure what adhoc meant exactly before checking the definition to be honest.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for your work on this @temyurchenko ! There's some warning in the CI job that need to be fixed before merging and we could rename "adhoc_module" if it makes sense, but LGTM.

**Comment by temyurchenko** [comment]
> Awesome! Great job working with us to get this merged. Working with you was a joy!
> 
> _I would be okay with adding a Changelog for this as I think you work should be recognized!_

Thank you, same for you and @Pierre-Sassoulas! 

I will do one final pass over the changes here, fix the commit message and fix the the linting errors. Not sure, if the changelog is needed, since this is mostly an internal change.

**Comment by temyurchenko** [comment]
> we could rename "adhoc_module" if it makes sense

Okay, I'll go with "synthetic_root".

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**
